s2/gen.go

```diff
@@ -0,0 +1,695 @@
+	Doc(name+" encodes a non-empty src to a guaranteed-large-enough dst.",
+		"It assumes that the varint-encoded length of the decompressed bytes has already been written.", "")
+	Pragma("noescape")
+
+	// "var table [maxTableSize]uint32" takes up 65536 bytes of stack space. An
+	// extra 56 bytes, to call other functions, and an extra 64 bytes, to spill
+	// local variables (registers) during calls gives 65536 + 56 + 64 = 65656.
+	var (
+		tableSize = 1 << tableBits
+		//tableMask  = tableSize - 1
+		baseStack  = 56
+		extraStack = 64
+		allocStack = baseStack + extraStack + tableSize
+	)
+	// Memzero needs at least 128 bytes.
+	if tableSize < 128 {
+		panic("tableSize must be at least 128 bytes")
+	}
+
+	lenSrcBasic, err := Param("src").Len().Resolve()
+	if err != nil {
+		panic(err)
+	}
+	lenSrc := lenSrcBasic.Addr
+
+	stack := AllocLocal(allocStack)
+	table := stack.Offset(allocStack - tableSize)
+
+	tmpStack := baseStack
+	// sLimit is when to stop looking for offset/length copies.
+	sLimit := stack.Offset(tmpStack)
+	tmpStack += 4
+	// Bail if we can't compress to at least this.
+	dstLimit := stack.Offset(tmpStack)
+	tmpStack += 4
+	nextEmit := stack.Offset(tmpStack)
+	tmpStack += 4
+	repeat := stack.Offset(tmpStack)
+	tmpStack += 4
+	dstBaseBasic, err := Param("dst").Base().Resolve()
+	if err != nil {
+		panic(err)
+	}
+	dstBase := dstBaseBasic.Addr
+
+	if tmpStack > extraStack+baseStack {
+		panic(fmt.Sprintf("tmp stack exceeded", tmpStack))
+	}
+	// Zero table
+	{
+		iReg := GP64()
+		MOVQ(U32(tableSize/8/16), iReg)
+		tablePtr := GP64()
+		LEAQ(table, tablePtr)
+		zeroXmm := XMM()
+		PXOR(zeroXmm, zeroXmm)
+
+		Label("zeroLoop" + name)
+		for i := 0; i < 8; i++ {
+			MOVOU(zeroXmm, Mem{Base: tablePtr, Disp: i * 16})
+		}
+		ADDQ(U8(16*8), tablePtr)
+		DECQ(iReg)
+		JNZ(LabelRef("zeroLoop" + name))
+
+		// nextEmit is where in src the next emitLiteral should start from.
+		MOVL(iReg.As32(), nextEmit)
+	}
+
+	{
+		const inputMargin = 8
+		tmp, tmp2, tmp3 := GP64(), GP64(), GP64()
+		MOVQ(lenSrc, tmp)
+		LEAQ(Mem{Base: tmp, Disp: -5}, tmp2)
+		// sLimit := len(src) - inputMargin
+		LEAQ(Mem{Base: tmp, Disp: -inputMargin}, tmp3)
+		// dstLimit := len(src) - len(src)>>5 - 5
+		SHRQ(U8(5), tmp)
+		SUBL(tmp2.As32(), tmp.As32())
+		MOVL(tmp3.As32(), sLimit)
+		MOVL(tmp.As32(), dstLimit)
+	}
+
+	// s = 1
+	s := GP64()
+	MOVB(U8(1), s.As8())
+	// repeat = 1
+	MOVL(s.As32(), repeat)
+
+	src := GP64()
+	Load(Param("src").Base(), src)
+
+	// Load cv
+	cv := GP64()
+	MOVQ(Mem{Base: src, Index: s, Scale: 1}, cv)
+	Label("mainLoop" + name)
+	{
+		Label("searchLoop" + name)
+		candidate := GP64()
+		{
+			nextS := GP64()
+			// nextS := s + (s-nextEmit)>>6 + 4
+			{
+				tmp := GP64()
+				MOVL(nextEmit, tmp.As32())
+				SUBL(s.As32(), tmp.As32())
+				SHRL(U8(skipLog), tmp.As32())
+				LEAQ(Mem{Base: s, Disp: 4, Index: tmp, Scale: 1}, nextS)
+			}
+			// if nextS > sLimit {goto emitRemainder}
+			// FIXME: failed to allocate registers???
```

Comment:
Just had a quick look, and I'm not immediately sure what's causing the allocation failure. The error message and tooling isn't as helpful as it needs to be.

However, a fairly simple fix for you in the short term is to hoist the `tmp*` registers out of the loop. For example have a line like

```go
	tmp, tmp2, tmp3 := GP64(), GP64(), GP64()
```

near the top of the function and use these instead of all the locally allocated temporaries.

Reply:
@mmcloughlin

Yeah, I see the workarounds, but it is more of a design choice to use scopes to limit the potential of "leaking" registers and create false dependencies and make the compiler more "free" to choose registers.

Reply:
@mmcloughlin Actually moving the `tmp := GP64()` out a level and having below `nextS := GP64()` and reusing the temp reg still yields the error. Very strange.